src/bug.cpp

```diff
@@ -0,0 +1,86 @@
+#include <memory>
+#include <vector>
+
+#include <simulator/bug.hpp>
+#include <simulator/map_cell.hpp>
+
+namespace simulation
+{
+  Bug::Bug(std::shared_ptr<MapCell> cell, const bool is_predator) : _cell(cell), _is_prey(!is_predator) {}
+
+  bool Bug::is_dead() const {
+    return !_is_alive;
+  }
+
+  bool Bug::is_prey() const {
+    return _is_prey;
+  }
+
+  void Bug::play_round() const {
+    
+  }
+
+  void Bug::get_eaten() {
+    _is_alive = false;
+    notify_death(_cell->get_bug());
+  }
+
+  uint8_t Bug::get_rounds_to_breed() const {
+    return DEFAULT_ROUNDS_TO_BREED;
+  }
+
+  uint8_t Bug::get_rounds_to_starve() const {
+    return DEFAULT_ROUNDS_TO_STARVE;
+  }
+
+  void Bug::reset_starve_counter() {
+    _starve_counter = get_rounds_to_starve();
+  }
+
+  void Bug::reset_breed_counter() {
+    _breed_counter = get_rounds_to_breed();
+  }
+
+  void Bug::increase_lifecycle_counters() {
+    _breed_counter++;
+    _starve_counter++;
+  }
+
+  void Bug::eat_move() {
+    for(auto adyacent : _cell->get_adyacents()) {
+      if(adyacent && adyacent->get_bug() && adyacent->get_bug()->is_prey()) {
+        notify_death(adyacent->get_bug());
+        reset_starve_counter();
+        adyacent->set_bug(_cell->get_bug());
```

Comment:
Yes, but otherwise we should create a shared pointer to `this`. So, this solution is he easiest way to pass a pointer to ourselves without making a more complex / less readable design.